examples/ecs/state.rs

```diff
@@ -6,26 +6,31 @@ use bevy::prelude::*;
 fn main() {
     App::new()
         .add_plugins(DefaultPlugins)
-        .add_state(AppState::Menu)
+        .add_state::<AppState>()
         .add_startup_system(setup)
-        .add_system_set(SystemSet::on_enter(AppState::Menu).with_system(setup_menu))
-        .add_system_set(SystemSet::on_update(AppState::Menu).with_system(menu))
-        .add_system_set(SystemSet::on_exit(AppState::Menu).with_system(cleanup_menu))
-        .add_system_set(SystemSet::on_enter(AppState::InGame).with_system(setup_game))
-        .add_system_set(
-            SystemSet::on_update(AppState::InGame)
-                .with_system(movement)
-                .with_system(change_color),
-        )
+        .add_system_to_schedule(setup_menu, &OnEnter(AppState::Menu))
+        .add_system(menu.on_update(AppState::Menu))
+        .add_system_to_schedule(cleanup_menu, &OnExit(AppState::Menu))
+        .add_system_to_schedule(setup_game, &OnEnter(AppState::InGame))
+        .add_systems((movement, change_color).on_update(AppState::InGame))
         .run();
 }
 
-#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash)]
 enum AppState {
+    #[default]
     Menu,
     InGame,
 }
 
+impl States for AppState {
+    type Iter = [AppState; 2];
+
+    fn variants() -> Self::Iter {
+        [AppState::Menu, AppState::InGame]
+    }
+}
```

Comment:
I have used the pattern of iterating through enum variants and a lot and very often I would add a new variant at some point, but forget to add it in the `variants` impl. In that case I believe the systems added to the new state variant would silently not run at all, I think.

Is the plan to add a derive `States` macro? (not for this PR of course).

Reply:
Yes, I'd very much like to add that in a separate PR.